rust/lance-encoding-datafusion/src/zone.rs

```diff
@@ -0,0 +1,211 @@
+// SPDX-License-Identifier: Apache-2.0
+// SPDX-FileCopyrightText: Copyright The Lance Authors
+
+use std::sync::Arc;
+
+use arrow_array::{ArrayRef, RecordBatch, UInt32Array};
+use arrow_schema::{Field, Schema};
+use datafusion_common::{arrow::datatypes::DataType, ScalarValue};
+use datafusion_expr::Accumulator;
+use datafusion_physical_expr::expressions::{MaxAccumulator, MinAccumulator};
+use futures::{future::BoxFuture, FutureExt};
+use lance_encoding::encoder::{
+    encode_batch, CoreFieldEncodingStrategy, EncodedMetadataBuffer, FieldEncoder,
+};
+
+use lance_core::Result;
+use lance_file::v2::writer::EncodedBatchWriteExt;
+
+struct CreatedZoneMap {
+    min: ScalarValue,
+    max: ScalarValue,
+    null_count: u32,
+}
+
+/// A field encoder that creates zone maps for the data it encodes
+///
+/// This encoder will create zone maps for the data it encodes.  The zone maps are created by
+/// dividing the data into zones of a fixed size and calculating the min/max values for each
+/// zone.  The zone maps are then encoded as metadata.
+///
+/// This metadata can be used by the reader to skip over zones that don't contain data that
+/// matches the query.
+pub struct ZoneMapsFieldEncoder {
+    items_encoder: Box<dyn FieldEncoder>,
+    items_type: DataType,
+
+    rows_per_map: u32,
+
+    maps: Vec<CreatedZoneMap>,
+    cur_offset: u32,
+    min: MinAccumulator,
+    max: MaxAccumulator,
+    null_count: u32,
+}
+
+impl ZoneMapsFieldEncoder {
+    pub fn try_new(
+        items_encoder: Box<dyn FieldEncoder>,
+        items_type: DataType,
+        rows_per_map: u32,
+    ) -> Result<Self> {
+        let min = MinAccumulator::try_new(&items_type)?;
+        let max = MaxAccumulator::try_new(&items_type)?;
+        Ok(Self {
+            rows_per_map,
+            items_encoder,
+            items_type,
+            min,
+            max,
+            null_count: 0,
+            cur_offset: 0,
+            maps: Vec::new(),
+        })
+    }
+}
+
+impl ZoneMapsFieldEncoder {
+    fn new_map(&mut self) -> Result<()> {
+        // TODO: We should be truncating the min/max values here
```

Comment:
Truncating binary/string values seems like a good idea. There are special considerations for the max to make sure you actually get a greater value after truncation.